modules/mqtt/testcontainers/mosquitto/__init__.py

```diff
@@ -0,0 +1,247 @@
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+import os
+import time
+from testcontainers.core.container import DockerContainer
+from testcontainers.core.waiting_utils import wait_container_is_ready
+from typing_extensions import Self
+
+from paho.mqtt import client as mqtt_client
+import paho.mqtt.enums
+from queue import Queue
+from typing import Optional
+
+# MosquittoContainer
+
+
+class MosquittoContainer(DockerContainer):
+    """
+    Specialization of DockerContainer for MQTT broker Mosquitto.
+    Example:
+
+        .. doctest::
+
+            >>> from testcontainers.mosquitto import MosquittoContainer
+
+            >>> with MosquittoContainer() as mosquitto_broker:
+            ...     mqtt_client = mosquitto_broker.get_client()
+    """
+
+    TESTCONTAINER_CLIENT_ID = "TESTCONTAINER-CLIENT"
+    DEFAULT_PORT = 1883
+    CONFIG_FILE = "integration-test-mosquitto.conf"
+
+    def __init__(
+        self,
+        image: str = "eclipse-mosquitto:latest",
+        port: int = None,
+        configfile: Optional[str] = None,
+        password: Optional[str] = None,
+        **kwargs,
+    ) -> None:
+        # raise_for_deprecated_parameter(kwargs, "port_to_expose", "port")
+        super().__init__(image, **kwargs)
+
+        if port is None:
+            self.port = MosquittoContainer.DEFAULT_PORT
+        else:
+            self.port = port
+        self.password = password
+
+        # setup container:
+        self.with_exposed_ports(self.port)
+        if configfile is None:
+            # default config ifle
+            TEST_DIR = os.path.dirname(os.path.abspath(__file__))
+            configfile = os.path.join(TEST_DIR, MosquittoContainer.CONFIG_FILE)
+        self.with_volume_mapping(configfile, "/mosquitto/config/mosquitto.conf")
+        if self.password:
+            # TODO: add authentication
+            pass
+
+        # helper used to turn asynchronous methods into synchronous:
+        self.msg_queue = Queue()
+
+        # dictionary of watched topics and their message counts:
+        self.watched_topics = {}
+
+        # reusable client context:
+        self.client = None
+
+    @wait_container_is_ready()
+    def _connect(self) -> None:
+        client, err = self.get_client()
+        if err != paho.mqtt.enums.MQTTErrorCode.MQTT_ERR_SUCCESS:
+            raise RuntimeError(f"Failed to estabilish a connection: {err}")
+
+        interval = 1.0
+        timeout = 5
+        start = time.time()
+        while True:
+            duration = time.time() - start
+            if client.is_connected():
+                return
+            if duration > timeout:
+                raise TimeoutError(f"Failed to estabilish a connection after {timeout:.3f} " "seconds")
```

Comment:
this is a double wait - you're supposed to throw from wait_container_is_ready i think

Reply:
right -- good catch. I've gone through the code again and I better understood how it works...

Reply:
resolving the conversation as I think I put a better implementation in place